_source/_posts/2019-09-19-nodejs-typescript.md

```diff
@@ -0,0 +1,637 @@
+
+## TypeScript to the rescue
+
+The key to making a large (or even small!) codebase readable is to keep track of all the data types flowing through the app.
+
+To help developers with this, TypeScript provided a typed superset of plain old JavaScript.
+This means that a regular JavaScript developer's existing knowledge applies to TypeScript.
+
+There are other efficient ways developers can use their knowledge and code in new ways. Developers who use NodeJS can now recycle their frontend code for the backend, and vice versa.
+
+Additionally, the introduction of static typing to JavaScript allows developers to reuse some of that old Java/C# knowledge to help maintain their NodeJS codebases.
+
+## Today's Project: A Simple Blogging Engine
+
+In this tutorial, we need a new backend for a blogging engine. This app will contain a list of blog posts, each with a title, content, and author.
+
+The authors will be able to register and log in using the session-in-cookies approach. Okta will manage the users, and all blog posts will be saved in memory.
+
+Documentation for the backend will be exposed as a Swagger specification generated from TypeScript typings for classes and methods.
+
+## Dependencies for the NodeJS Typescript Sample App
+
+- NestJS to bootstrap the server
+- Swagger to define the server docs
+- Okta for user management and authentication
+- NodeJS
+- `validate-class` npm library for input validation
+- nodemon to auto-restart the server on every change
+
+## Initialize
+
+Create an empty directory `bloggeur`, `cd` into it, and type `npm init` there. After pressing the `ENTER` key  a few times, a NodeJS project will initiate.
+
+## TypeScript config JSON
+
+As TypeScript is very configurable, most TypeScript projects contain a configuration to select the language features it should use.
+
+Create a `tsconfig.json` in the root of your project:
+
+```json
+{
+  "compilerOptions": {
+    "target": "es2015",
+    "lib": ["es2015"],
+    "types": ["reflect-metadata", "node"],
+    "module": "commonjs",
+    "moduleResolution": "node",
+    "experimentalDecorators":true,
+    "emitDecoratorMetadata": true,
+    "sourceMap": true,
+    "declaration": false,
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true
+  },
+  "exclude": [
+    "node_modules"
+  ]
+}
+```
+
+## Add all the dependencies
+
+```bash
+npm install @nestjs/common@6.5.2 @nestjs/core@6.5.2 @nestjs/platform-express@6.5.2 @nestjs/swagger@3.1.0 @okta/okta-auth-js@2.6.0 @okta/okta-sdk-nodejs@2.0.0 @types/express@4.17.0 @types/node@12.6.2 class-transformer@0.2.3 class-validator@0.9.1 cookie-parser@1.4.4 dotenv@8.0.0 install@0.13.0 reflect-metadata@0.1.13 rxjs@6.5.2 swagger-ui-express@4.0.7 ts-node@8.3.0 typescript@3.5.3 glob@7.1.4 @types/glob@7.1.1
```

Comment:
Wow - that's a lot of dependencies! Maybe break it up into several lines so it doesn't hide most of it?